### ISSUE TITLE
Bridging to VPN interface and `en0` removes SSH access

### ISSUE THREAD

**Issue by fmoc** [cirruslabs/tart#473]
I have a setup on a Mac mini M1 which is connected to the Internet via LAN (`en0`) and to a VPN interface (`utun0` resp. `bridge100`). I am trying to set up networking without the [NAT workaround mentioned here](https://github.com/cirruslabs/tart/issues/384).

```sh
# first attempt using the bridge
> tart run ... --net-bridged bridge100
Error: no bridge interfaces matched "bridge100", available interfaces: en0 (or "Ethernet"), en1 (or "Wi-Fi")

# starts up, but doesn't allow me to SSH into the VM using the IP address provided by tart ip
> tart run ... --net-bridged utun0 --net-bridged en0
```

The latter is not an option, since I permanently need access via SSH to the VM for monitoring purposes.

How do I need to run the VM to have it connect to both networks as well as make sure I can access it from the host?

**Comment by fkorotkov** [comment]
Just to clarify:

1. You need VPN access inside your VM to access some private stuff.
2. You need local LAN bridge so you can SSH into the VM from a different host.

Did I understand your setup correctly? In that case I think  we can an option to provide multiple interfaces for the bridged mode since [`VZVirtualMachineConfiguration` supports it](https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656724-networkdevices).

Can I also ask you to provide output of `pfctl -s nat -a "com.apple.internet-sharing/shared_v4"` command executed on the host where you run VMs.

**Comment by edigaryev** [comment]
> I'm not sure what to bridge to to have access from the host

Does the standard shared network (NAT) allows access from the host for you? Or I'm misunderstanding the notion of host?

The same shared network would also allow access to VPN resources, if the VPN routes are configured adequately. You can check this with `pfctl -s nat -a "com.apple.internet-sharing/shared_v4"`, as @fkorotkov mentioned above.

**Comment by fkorotkov** [comment]
In the meantime I'm adding support for Bridged network to Orchard in https://github.com/cirruslabs/orchard/pull/78.

In case of Orchard, just having one bridged network for VPN will be enough since SSHing is done through a runnel via an agent sitting on the host.

**Comment by fmoc** [comment]
> Does the standard shared network (NAT) allows access from the host for you? Or I'm misunderstanding the notion of host?

Yes, that works fine. However, without the added NAT rule mentioned in https://github.com/cirruslabs/tart/issues/384#issuecomment-1406196317, I cannot access the VPN interface `utun*`. The current rule set with the NAT rule (which I kind of need to hack in on demand, currently with a cronjob that runs every minute) looks like this:

```sh
> sudo pfctl -s nat -a com.apple.internet-sharing/shared_v4
No ALTQ support in kernel
ALTQ related functions disabled
nat on en0 inet from 192.168.64.0/24 to any -> (en0:0) extfilter ei
no nat on bridge100 inet from 192.168.64.1 to 192.168.64.0/24
nat on utun0 inet from 192.168.64.0/24 to any -> 172.XX.YY.ZZ  # VPN
```

The latter rule is the one I add using that cron job, which basically adds the following rule:

```
nat on utun0 from bridge100:network to any -> utun0
```

This rule is translated to the syntax seen above.

> if the VPN routes are configured adequately

I think tart should probably receive a configuration feature or CLI parameter where you can specify the VPN interface so it can automatically add that rule. I'm sure I'm not the only person in this community that depends on such a setup.

**Comment by edigaryev** [comment]
>I think tart should probably receive a configuration feature or CLI parameter where you can specify the VPN interface so it can automatically add that rule. I'm sure I'm not the only person in this community that depends on such a setup.

I'm afraid that this would be too much of a hack to implement since this requires `root` and messes up with Apple's PF tables.

I think a more productive direction would be figure out why in your case the NAT rules are not getting added automatically.

For example, when I run the Tailscale VPN, the needed PF rules are inserted automagically:

```
% sudo pfctl -s nat -a com.apple.internet-sharing/shared_v4 2>/dev/null | grep utun
nat on utun5 inet from 192.168.62.0/24 to any -> (utun5:0) extfilter ei
```

As a result, both LAN and VPN networks are accessible from the guests when running default shared (NAT) networking (no `--net-softnet` or `--net-bridged`).

Perhaps you could provide some ways to reproduce your case? And/or share some more details on your VPN configuration, like the VPN client used, the routes installed (`netstat -rn -f inet | grep utun`) and if the VPN connection is visible in `Settings → Network`?

**Comment by fmoc** [comment]
We use OpenVPN (directly, via homebrew, not via Tunnelblick).

![screenshot_2023-04-20_14-48-30](https://user-images.githubusercontent.com/80399010/233371118-647f63f4-c789-456b-bea1-1d8978c10f86.png)

```
> % sudo netstat -rn -f inet
Routing tables

Internet:
Destination        Gateway            Flags           Netif Expire
default            wan.1              UGScg             en0       
default            link#18            UCSIg       bridge100      !
10.90.4/24         v.p.n.1            UGSc            utun0       
10.90.5/24         v.p.n.1            UGSc            utun0       
127                127.0.0.1          UCS               lo0       
127.0.0.1          127.0.0.1          UH                lo0       
wan/26             link#6             UCS               en0      !
wan.1/32           link#6             UCS               en0      !
wan.1              8:5:e2:c8:93:ab    UHLWIir           en0    428
wan.7/32           link#6             UCS               en0      !
wan.7              14:98:77:84:46:57  UHLWIi            lo0       
169.254            link#6             UCS               en0      !
v.p                v.p.n.1            UGSc            utun0       
v.p.n/22           v.p.n.31           UGSc            utun0       
v.p.n.31           v.p.n.31           UH              utun0       
192.168.64         link#18            UC          bridge100      !
192.168.64.38      5a.32.5a.f9.53.5   UHLWIi      bridge100    498
192.168.64.39      fa.6f.af.8c.a4.bb  UHLWIi      bridge100    534
224.0.0/4          link#6             UmCS              en0      !
224.0.0.251        1:0:5e:0:0:fb      UHmLWI            en0       
255.255.255.255/32 link#6             UCS               en0      !
```

**Comment by edigaryev** [comment]
Will check if allowing the `tart run` to take multiple `--net-bridged` flags (and thus adding multiple virtual Ethernet adapters to the guest) yields any positive results and will post my findings here.

**Comment by edigaryev** [comment]
I've just realized that `--net-bridged utun*` would never work, because it's not a bridged interface.

Neither manually creating a bridged interface with `ifconfig` would work, it simply won't be added to the `VZBridgedNetworkInterface.networkInterfaces` for some reason.

Unfortunately, at this point I don't see any solid options besides switching to a VPN solution that uses macOS APIs (e.g. WireGuard), which will guarantee proper NAT functioning.

As a side note, regarding the ability to specify multiple interfaces for a VM: it is possible to implement, but is somewhat complicated because we'd need to have more than one MAC-address for a VM.

If we re-use the single MAC from the VM's config, `Virtualization.Framework` will de-duplicate all the interfaces we pass to it internally based on MAC-address and the VM will run with only a single interface.

**Comment by fmoc** [comment]
Thanks! I would have never guessed that using some native macOS API to set up VPN in the UI, too, would make such a big difference...

> I've just realized that `--net-bridged utun*` would never work, because it's not a bridged interface.

The interface doesn't have to be a bridge to be added to a bridge, right?

**Comment by edigaryev** [comment]
> The interface doesn't have to be a bridge to be added to a bridge, right?

You're right, however, `Virtualization.Framework`'s [`VZBridgedNetworkInterface`](https://developer.apple.com/documentation/virtualization/vzbridgednetworkinterface) only allows us to use ["network interfaces available for bridging"](https://developer.apple.com/documentation/virtualization/vzbridgednetworkinterface/3656608-networkinterfaces), which is unfortunate.

This definition seems to only include (1) physical interfaces like Wi-Fi and Ethernet, attachable adapters (e.g. USB to Ethernet) and (2) virtual interfaces created from the Apple's UI or using networking APIs.

**Comment by fmoc** [comment]
Yup, makes sense. Thanks for all the debugging, I hope this is going to help others, too.

**Comment by fkorotkov** [comment]
Interesting find, @Sn0wCrack! Thank you for sharing! We'll investigate.

**Comment by edigaryev** [comment]
Hello @Sn0wCrack, this is interesting indeed!

1. Which version of the [OpenVPN Connect](https://openvpn.net/client/) did you use? Or was it some other OpenVPN client?
2. Which version of VMWare Fusion did you use?
3. Was the VMWare Fusion idle at the time you've observed this, or some VM was running in it?
4. Which OS and network settings did you use for that VMWare Fusion VM?

Currently, as far as I'm aware, the only way to achieve something like that is to interact with the Packet Filter (PF) firewall through `/dev/pf` directly or via `pfctl`. Some nice libraries exist, that allow you to call [`add_nat_rule()`](https://docs.rs/pfctl/latest/pfctl/struct.PfCtl.html#method.add_nat_rule) to achieve that.

However, the main issue with this approach is that it's hard to guarantee that a PF rule that we've insterted will be removed when `tart run` fails to terminate cleanly.

**Comment by edigaryev** [comment]
@Sn0wCrack  was not able to reproduce the same using OpenVPN Connect 3.8.1 and VMWare Fusion 25H2 on macOS 26.2 (Tahoe).

In fact, when the VMWare Fusion VM is running, there are no NAT entries created at all:

```
% sudo pfctl -s nat -a "com.apple.internet-sharing/shared_v4"
No ALTQ support in kernel
ALTQ related functions disabled
```

But the VMWare Fusion VM can indeed access the networks exposed through the OpenVPN interface.

**Comment by Sn0wCrack** [comment]
Very strange. I can attest that having VMware Fusion installed was the difference between this not working and working for myself and a colleague.

Additionally as mentioned we also had installed the VMware Fusion Vagrant utility and have been launching the Tart VM via Vagrant.

We also additionally have firewall rules enabled on our Macs as part of our company's MDM policies.

One thing I noticed earlier in my testing was that toggling the Firewall on and off (or changing any firewall settings) seemed to add and removing additional vmenet and bridge networks to appear in the output of ifconfig. However after a reboot this stopped occurring.

Later in our testing I reinstalled VMware Fusion after noting it was the only difference  between myself and my colleage. After doing so the packet forwarding rules and interfaces were consistently added without touching any firewall related setting.
